### ISSUE TITLE
Incorrect R Peak Detection

### ISSUE THREAD

**Issue by derek-nakano** [luishowell/ecg-detectors#2]
Hi Luis, I'm having issues with the R peak detection for almost every detector.  The detectors detect locations close to the R peak, but not exactly.  Here is a snippet of code I took from the example you created for Amri95's issue and get different results.
```python
sig, fields = wfdb.rdsamp(r'test_data\mit-bih-normal-sinus-rhythm-database-1.0.0\16265')
unfiltered_ecg = sig[:3000, 0]
fs = 128

detectors = Detectors(fs)
# r_peaks = detectors.two_average_detector(unfiltered_ecg)
#r_peaks = detectors.matched_filter_detector(unfiltered_ecg)
#r_peaks = detectors.swt_detector(unfiltered_ecg)
#r_peaks = detectors.engzee_detector(unfiltered_ecg)
r_peaks = detectors.christov_detector(unfiltered_ecg)
#r_peaks = detectors.hamilton_detector(unfiltered_ecg)
# r_peaks = detectors.pan_tompkins_detector(unfiltered_ecg)

plt.figure(figsize=(20,10))
plt.plot(unfiltered_ecg)
plt.plot(r_peaks, unfiltered_ecg[r_peaks], 'ro')
plt.title('Detected R-peaks')
plt.show()
```
![Capture](https://user-images.githubusercontent.com/59340088/71560058-dbe52d00-2a19-11ea-9b1a-d63a7befbea8.PNG)

**Comment by luishowell** [comment]
Hi, in the most recent commits there was an attempt to make the detectors more "realistic" from a real-time perspective, however, this had a serious impact on performance which is what you are seeing above. I have rolled back the master branch to a previous commit where the performance is optimal and move the more "realistic" code to the with-delay branch. Pull the master branch and run your code again. If you are still having issues, delete your local repo and clone it again.

**Comment by berndporr** [comment]
Most ECG R peak detectors don't detect the peak so the name is a misnomer but used anyway. Most detectors work with a threshold. Only a few do actually a maximum detection and that's usually dangerous because a peak might have multiple peaks in real life. The earlier version first used the detector and then looked backwards by detecting the maximum. This however then measured not the performance of the R peak detector but the performance of the Python function searching for the peak -- in this case `np.argmax`.
In other words most detectors cause a delay between R peak and its detection. That should be of course constant.

**Comment by berndporr** [comment]
[This](https://github.com/berndporr/py-ecg-detectors) is the code we used to publish the [preprint](https://www.biorxiv.org/content/10.1101/722397v3). In that the delay introduced by most detectors was measured by taking the median of the delays and then subtracting it.

**Comment by berndporr** [comment]
I just had a look and then detectors which do peak detection with `argmax` are: EngZee, TwoAverage and Pan Tompkins. The others return the R peaks with a delay.

**Comment by derek-nakano** [comment]
Hi Luis, the detectors are locating the R peaks accurately now, thanks!

Hi Bern, I tested out the code you linked.  This is what I'm getting for the R peaks for the detectors you mentioned:
![Capture](https://user-images.githubusercontent.com/59340088/71592578-a73ea780-2ae5-11ea-8878-72b3ea7afc64.PNG)

From the image, it seems like Pan Tompkins and Two Average have an R peak delay while EngZee does not.

**Comment by berndporr** [comment]
```
                if detection[peak]>threshold_I1 and (peak-signal_peaks[-1])>0.3*fs:                       
                    signal_peaks.append(peak)
```
Ah yes. You are right. I did only a quick `grep` to see which one uses `argmax` but PT actually uses two thresholds.

**Comment by berndporr** [comment]
Note that "locating R peaks accurately" is dangerous to say because in noisy ECGs there is rarely just one peak per QRS complex. For that reason detectors use thresholds where the velocity of the filtered signal is fastest so that noise has least impact. Plus adding `argmax` after every detector is **dangerous/misleading** because it turns them all into one **identical** detector where `argmax` does the detection -- and the actual timestamp of the detector is discarded. That was a mistake in the earlier version and was fixed in the py-ecg-detectors.

**Comment by berndporr** [comment]
BTW: we had an interesting discussion recently on twitter about measuring very noisy ECG between the ears (!) of a person and the two average detector worked best.

**Comment by derek-nakano** [comment]
I see, so that's why most detectors don't detect the R-peaks.  I was under the impression that Pan and Tompkins was the best algorithm in detecting R-peaks, but after reading your article, it has one of the worst performances of the algorithms tested.  So many detectors claim a high sensitivity, but they conveniently omit the tolerance.  Do you have a link to the twitter discussion?  I'm interested in reading more about this topic.

**Comment by berndporr** [comment]
I have linked one discussion here:
https://github.com/berndporr/hrv

**Comment by derek-nakano** [comment]
Okay, thanks!
